MBVM-88: Fetch DB dumps via HTTP instead of FTP

atj opened the pull request:
User have reported issues fetching DB dumps using FTP, and further investigation revealed that they were due to the use of IPv4-mapped IPv6 addresses. Unfortunately this issue cannot be resolved easily due to the idiosyncratic nature of the FTP protocol. This commit changes the default download protocol from FTP to HTTP, which will resolve the issue and allow us to make DB dumps available over IPv6 in future.

Note that variable names containing FTP have not been changed for backward compatibility reasons.

atj committed:
Fetch DB dumps via HTTP instead of FTP

User have reported issues fetching DB dumps using FTP, and further
investigation revealed that they were due to the use of IPv4-mapped IPv6
addresses. Unfortunately this issue cannot be resolved easily due to the
idiosyncratic nature of the FTP protocol.

This commit adds a new `MUSICBRAINZ_MIRROR_URL` variable with a default
value of `http://ftp.eu.metabrainz.org/pub/musicbrainz`, which will
resolve the issue and allow us to make DB dumps available over IPv6 in
future.

For backward compatibility reasons the `MUSICBRAINZ_BASE_FTP_URL`
variable is still supported and overrides the new
`MUSICBRAINZ_MIRROR_URL` if set.

atj commented:
This is the only test I've run:

```
$ docker-compose run --rm musicbrainz createdb.sh -sample -fetch
[+] Running 6/6
 ✔ Network musicbrainz-docker_default        Created                                                                                                                                                                                     0.1s 
 ✔ Container musicbrainz-docker-validator-1  Created                                                                                                                                                                                     0.2s 
 ✔ Container musicbrainz-docker-db-1         Created                                                                                                                                                                                     0.2s 
 ✔ Container musicbrainz-docker-redis-1      Created                                                                                                                                                                                     0.2s 
 ✔ Container musicbrainz-docker-mq-1         Created                                                                                                                                                                                     0.2s 
 ✔ Container musicbrainz-docker-search-1     Created                                                                                                                                                                                     0.2s 
[+] Running 5/5
 ✔ Container musicbrainz-docker-mq-1         Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-db-1         Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-validator-1  Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-redis-1      Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-search-1     Started                                                                                                                                                                                     0.9s 
Mon 17 Apr 2023 12:16:54 PM UTC: Fetching database dump...
--2023-04-17 12:16:54--  http://ftp.eu.metabrainz.org/pub/musicbrainz/data/sample/LATEST
Resolving ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)... 138.201.203.43
Connecting to ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)|138.201.203.43|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16 [application/octet-stream]
Saving to: ‘/media/dbdump/LATEST’

LATEST                                                      100%[=========================================================================================================================================>]      16  --.-KB/s    in 0.008s  

2023-04-17 12:16:57 (1.85 KB/s) - ‘/media/dbdump/LATEST’ saved [16/16]

--2023-04-17 12:16:57--  http://ftp.eu.metabrainz.org/pub/musicbrainz/data/sample/20230401-000001/mbdump-sample.tar.xz
Resolving ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)... 138.201.203.43
Connecting to ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)|138.201.203.43|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 286747812 (273M) [application/octet-stream]
Saving to: ‘/media/dbdump/mbdump-sample.tar.xz’

mbdump-sample.tar.xz                                        100%[=========================================================================================================================================>] 273.46M  36.5MB/s    in 6.4s    

2023-04-17 12:17:04 (42.9 MB/s) - ‘/media/dbdump/mbdump-sample.tar.xz’ saved [286747812/286747812]

Mon 17 Apr 2023 12:17:04 PM UTC: Done fetching dump files.
```

I'm not sure about the changes to `fetch-dump.sh`, is anyone likely to be using a custom FTP server?

yvanzo commented on README.md:
```diff
@@ -341,16 +341,20 @@ By default, MusicBrainz Server uses 10 `plackup` processes at once.
 
-See also the [list of mirrors](https://musicbrainz.org/doc/MusicBrainz_Database/Download#Download);
-Only FTP URLs are supported here.
+An FTP mirror can be used instead by specifying the `MUSICBRAINZ_BASE_FTP_URL`
```
It seems that you rather just dropped `MUSICBRAINZ_BASE_FTP_URL` from the code.

yvanzo commented on README.md:
```diff
@@ -341,16 +341,20 @@ By default, MusicBrainz Server uses 10 `plackup` processes at once.
-Only FTP URLs are supported here.
+An FTP mirror can be used instead by specifying the `MUSICBRAINZ_BASE_FTP_URL`
+Docker environment variable. Note that support for FTP mirrors is deprecated
+and will be removed in a future release.
```
The option `--base-ftp-url` might be deprecated, but not the support for FTP as there are FTP mirrors.

For example we could replace both options with `--base-url` later on and support FTP/HTTP/RSYNC.

yvanzo commented on build/musicbrainz-dev/scripts/fetch-dump.sh:
```diff
@@ -21,8 +22,10 @@ Targets:
+  --base-ftp-url <url>          Specify URL to MetaBrainz/MusicBrainz FTP server.
+                                (Note: this option is deprecated and will be removed in a future release)
+  --base-http-url <url>         Specify URL to MetaBrainz/MusicBrainz FTP server.
+                                (Default: '$BASE_HTTP_URL')
```
Oversight
```suggestion
  --base-http-url <url>         Specify URL to MetaBrainz/MusicBrainz HTTP server.
                                (Default: '$BASE_HTTP_URL')
```

yvanzo commented on build/musicbrainz/scripts/fetch-dump.sh:
```diff
@@ -21,8 +22,10 @@ Targets:
+  --base-ftp-url <url>          Specify URL to MetaBrainz/MusicBrainz FTP server.
+                                (Note: this option is deprecated and will be removed in a future release)
+  --base-http-url <url>         Specify URL to MetaBrainz/MusicBrainz FTP server.
+                                (Default: '$BASE_HTTP_URL')
```
Same
```suggestion
  --base-http-url <url>         Specify URL to MetaBrainz/MusicBrainz HTTP server.
                                (Default: '$BASE_HTTP_URL')
```

yvanzo reviewed (commented):
Just a quick review without testing it so far.

atj commented:
Here's my limited testing so far:

* Checked defaults:

```
❯ docker-compose run --rm musicbrainz createdb.sh -sample -fetch
[+] Running 5/5
 ✔ Container musicbrainz-docker-redis-1      Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-mq-1         Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-validator-1  Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-db-1         Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-search-1     Created                                                                                                                                                                                     0.5s 
[+] Running 5/5
 ✔ Container musicbrainz-docker-mq-1         Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-validator-1  Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-redis-1      Started                                                                                                                                                                                     1.0s 
 ✔ Container musicbrainz-docker-db-1         Started                                                                                                                                                                                     1.0s 
 ✔ Container musicbrainz-docker-search-1     Started                                                                                                                                                                                     1.0s 
Mon 17 Apr 2023 04:57:34 PM UTC: Fetching database dump...
--2023-04-17 16:57:34--  http://ftp.eu.metabrainz.org/pub/musicbrainz/data/sample/LATEST
Resolving ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)... 138.201.203.43
Connecting to ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)|138.201.203.43|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 16 [application/octet-stream]
Saving to: ‘/media/dbdump/LATEST’

LATEST                                                      100%[=========================================================================================================================================>]      16  --.-KB/s    in 0s      

2023-04-17 16:57:34 (6.20 MB/s) - ‘/media/dbdump/LATEST’ saved [16/16]
```

* Checking that setting `MUSICBRAINZ_BASE_FTP_URL` envvar overrides `MUSICBRAINZ_MIRROR_URL`:

```
❯ export MUSICBRAINZ_BASE_FTP_URL=ftp://ftp.eu.metabrainz.org/pub/musicbrainz 
❯ docker-compose run --rm musicbrainz createdb.sh -sample -fetch
[+] Running 5/5
 ✔ Container musicbrainz-docker-validator-1  Created                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-mq-1         Created                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-db-1         Created                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-redis-1      Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-search-1     Created                                                                                                                                                                                     0.6s 
[+] Running 5/5
 ✔ Container musicbrainz-docker-db-1         Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-redis-1      Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-search-1     Started                                                                                                                                                                                     1.0s 
 ✔ Container musicbrainz-docker-validator-1  Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-mq-1         Started                                                                                                                                                                                     0.9s 
Mon 17 Apr 2023 05:01:04 PM UTC: Fetching database dump...
--2023-04-17 17:01:04--  ftp://ftp.eu.metabrainz.org/pub/musicbrainz/data/sample/LATEST
           => ‘/media/dbdump/LATEST’
Resolving ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)... 138.201.203.43
Connecting to ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)|138.201.203.43|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/musicbrainz/data/sample ... done.
==> SIZE LATEST ... 16
==> PASV ... done.    ==> RETR LATEST ... done.
Length: 16 (unauthoritative)

LATEST                                                      100%[=========================================================================================================================================>]      16  --.-KB/s    in 0.001s  

2023-04-17 17:01:05 (25.7 KB/s) - ‘/media/dbdump/LATEST’ saved [16]
```

* Checked that defining `MUSICBRAINZ_MIRROR_URL` envvar to an FTP URL works:

```
❯ export MUSICBRAINZ_MIRROR_URL=ftp://ftp.eu.metabrainz.org/pub/musicbrainz
❯ docker-compose run --rm musicbrainz createdb.sh -sample -fetch
[+] Running 5/5
 ✔ Container musicbrainz-docker-db-1         Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-redis-1      Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-mq-1         Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-validator-1  Created                                                                                                                                                                                     0.5s 
 ✔ Container musicbrainz-docker-search-1     Created                                                                                                                                                                                     0.5s 
[+] Running 5/5
 ✔ Container musicbrainz-docker-search-1     Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-validator-1  Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-db-1         Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-redis-1      Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-mq-1         Started                                                                                                                                                                                     0.6s 
Mon 17 Apr 2023 05:03:25 PM UTC: Fetching database dump...
--2023-04-17 17:03:25--  ftp://ftp.eu.metabrainz.org/pub/musicbrainz/data/sample/LATEST
           => ‘/media/dbdump/LATEST’
Resolving ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)... 138.201.203.43
Connecting to ftp.eu.metabrainz.org (ftp.eu.metabrainz.org)|138.201.203.43|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/musicbrainz/data/sample ... done.
==> SIZE LATEST ... 16
==> PASV ... done.    ==> RETR LATEST ... done.
Length: 16 (unauthoritative)

LATEST                                                      100%[=========================================================================================================================================>]      16  --.-KB/s    in 0.001s  

2023-04-17 17:03:25 (24.2 KB/s) - ‘/media/dbdump/LATEST’ saved [16]
```

* Checked that defining `MUSICBRAINZ_MIRROR_URL` envvar to a value without a protocol fails:

```
❯ export MUSICBRAINZ_MIRROR_URL=ftp.eu.metabrainz.org/pub/musicbrainz
❯ docker-compose run --rm musicbrainz createdb.sh -sample -fetch
[+] Running 5/5
 ✔ Container musicbrainz-docker-redis-1      Created                                                                                                                                                                                     0.4s 
 ✔ Container musicbrainz-docker-search-1     Created                                                                                                                                                                                     0.4s 
 ✔ Container musicbrainz-docker-db-1         Created                                                                                                                                                                                     0.4s 
 ✔ Container musicbrainz-docker-mq-1         Created                                                                                                                                                                                     0.4s 
 ✔ Container musicbrainz-docker-validator-1  Created                                                                                                                                                                                     0.4s 
[+] Running 5/5
 ✔ Container musicbrainz-docker-validator-1  Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-mq-1         Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-search-1     Started                                                                                                                                                                                     0.9s 
 ✔ Container musicbrainz-docker-redis-1      Started                                                                                                                                                                                     0.6s 
 ✔ Container musicbrainz-docker-db-1         Started                                                                                                                                                                                     0.6s 
[+] Building 2.2s (13/13) FINISHED                                                                                                                                                                                                            
fetch-dump.sh: --mirror-url must begin with ftp://, http:// or https://
```

yvanzo committed:
Refactor: Use "base download" instead of "mirror"

To avoid confusion with MusicBrainz mirror server which can be set up
using this Docker Compose project, use “(base) download” instead
“mirror” when it comes to mirrors serving data dump files.

yvanzo committed:
Reorder in alphabetical order after 127f5e23

yvanzo committed:
Amend 63363d2c: Fix a bug to match URL protocol

Skip escaping special characters in the regexp for `[[` bash built-in.

yvanzo reviewed (approved):
Thanks a lot @atj!
(Mainly discussed on IRC.)
I tested it too, with the amending commits.

atj commented:
Thanks for your help and feedback on this @yvanzo!
